### PR TITLE
Tagging

### PR DESCRIPTION
**Pull request by brtsmrtn**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -4,32 +4,82 @@ import { SubmitLink } from "./components/SubmitLink";
...
+      newTag = {
+        id: tagId.toString(),
+        title: title,
+        color: tagColor,
+      };
+      setKnownTags([...knownTags, newTag]);
+    }
+    return newTag;
+  };
+  const addTag = (listItem: ListItem, tag: Tag) => {
```
Comment (erikvanek) [inline src/App.tsx]: maybe `assignTagToItem` instead of `addTag` would suit better?

File: src/App.tsx [hunk 2]
```diff
@@ -4,32 +4,82 @@ import { SubmitLink } from "./components/SubmitLink";
...
+      tags: [...listItem.tags, tag],
+    };
+    newItems.splice(
+      newItems.findIndex((it) => it.id === newListItem.id),
+      1,
+      newListItem
+    );
+    setItems([...newItems]);
+  };
+  const deleteTag = (listItem: ListItem, tagItem: Tag) => {
```
Comment (erikvanek) [inline src/App.tsx]: `deleteTag` -> `unassignTag`?

File: src/App.tsx [hunk 3]
```diff
@@ -4,32 +4,82 @@ import { SubmitLink } from "./components/SubmitLink";
...
+    const newItems = [...items];
+    const newListItem = { ...listItem, tags: [...listItem.tags] };
+    newListItem.tags.splice(
+      newListItem.tags.findIndex((li) => li.id === tagItem.id),
```
Comment (erikvanek) [inline src/App.tsx]: Id change `(li)` to either `(item)` or `(listItem)`. `li` doesn't mean anything and is a step backwards [towards Orban](https://en.wikipedia.org/wiki/Hungarian_notation).

File: src/App.tsx [hunk 4]
```diff
@@ -4,32 +4,82 @@ import { SubmitLink } from "./components/SubmitLink";
...
+      1
+    );
+    newItems.splice(
+      newItems.findIndex((it) => it.id === newListItem.id),
```
Comment (erikvanek) [inline src/App.tsx]: `it` -> `item` imo.

File: src/App.tsx [hunk 5]
```diff
@@ -4,32 +4,82 @@ import { SubmitLink } from "./components/SubmitLink";
...
+      newListItem
+    );
+    setItems([...newItems]);
   };
   const [url, setUrl] = useState("");
   const [items, setItems] = useState<ListItem[]>([]);
+  const [knownTags, setKnownTags] = useState<Tag[]>([]);
   return (
     <div>
       <AddLink url={url.toLowerCase()} onChange={setUrl} />
```
Comment (erikvanek) [inline src/App.tsx]: Not part of this PR but I believe that the `toLowerCase()` call should hapen within `<AddLink .../>` instead of being called like this. Not a big deal but I think it's a bit cleaner that way.

File: src/components/Listing.tsx [hunk 6]
```diff
@@ -1,37 +1,52 @@
...
+};
+export const Listing: (props: ListingProps) => JSX.Element = ({
+  items,
+  knownTags,
+  title,
+  itemChanged,
+  tagAdded,
+  tagDeleted,
+  updateKnownTags,
+}: ListingProps) => {
```
Comment (erikvanek) [inline src/components/Listing.tsx]: Could be simplified even more if you want to.
E.g.
`const foo = () => { return (<text>Bar</text>) }` equals to `const foo = () => (<text>Bar</text>)`. See implicit return part of [this article](https://medium.com/@bunlong/arrow-functions-return-rules-in-javascript-b63ed5f25994) or [this one](https://waylonwalker.com/explicit-vs-implicit-returns-in-javascript).

File: src/components/SubmitLink.tsx [hunk 7]
```diff
@@ -11,21 +11,21 @@ type SubmitProps = UrlType & {
...
+  items,
+  onClick,
+  url,
+}: SubmitProps) => {
+  return (
+    <Button
+      variant="outlined"
+      color="primary"
+      disabled={
+        !isLinkValid(url) || isLinkDuplicate(url, items) || url.length >= 2083
```
Comment (erikvanek) [inline src/components/SubmitLink.tsx]: Some magic number appears here :)

File: src/components/TagButton.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  tagDeleted,
+  updateKnownTags,
+}: TagButtonProps) => {
+  const [dialogOpen, setDialogOpen] = useState(false);
+  const [errorTag, setErrorTag] = useState("");
+  const isTagNew = function (tagTitle: string, tags: Tag[]): Tag | undefined {
+    return tags.find((t) => t.title === tagTitle);
+  };
+  const onChangeAutocomplete = (
+    value: Tag[],
```
Comment (erikvanek) [inline src/components/TagButton.tsx]: I think the `value` parameter should be called `tag`.

File: src/components/TagButton.tsx [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    reason: string,
+    item: ListItem
+  ) => {
+    const targetTag = value[value.length - 1];
+    if (reason === "select-option") {
+      const tagAlreadyAssigned = isTagNew(targetTag.title, item.tags);
+      if (tagAlreadyAssigned) {
+        () => setErrorTag(errorTags.submit);
```
Comment (erikvanek) [inline src/components/TagButton.tsx]:
```suggestion
        setErrorTag(errorTags.submit);
```
not sure but I think we just want to call the function here.

File: src/components/TagButton.tsx [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        setErrorTag(errorTags.empty);
+      }
+    }
+  };
+  const onChangeTextField = (
+    e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>,
+    item: ListItem
+  ) => {
+    const targetTag = e.target as HTMLTextAreaElement;
+    const targetTagTitle = targetTag.value;
```
Comment (erikvanek) [inline src/components/TagButton.tsx]: `targetTagTitle` feels a bit clumsy. Can we do better than that?

File: src/components/TagButton.tsx [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            Add new tag or pick any from the list.
+          </DialogContentText>
+          <DialogContentText
+            style={{
+              fontStyle: "italic",
+              fontSize: 12,
+              marginTop: "20px",
+            }}
+          >
+            {errorTag ? errorTag : <br />}
```
Comment (erikvanek) [inline src/components/TagButton.tsx]: is `<br />` really needed?

File: src/components/TagList.tsx [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import React from "react";
+import Chip from "@material-ui/core/Chip";
+import { Tag } from "../types/Tag";
+import { ListItem } from "../types/ListItem";
+export type TagListProps = {
+  editable: boolean;
+  item: ListItem;
+  tagDeleted: (item: ListItem, tag: Tag) => void;
+};
+export const TagList: (props: TagListProps) => JSX.Element = ({
```
Comment (erikvanek) [inline src/components/TagList.tsx]: This one could also leverage implicit function return. Rule of thumb is as long as there is no logic within the `render` method, implicit return might be used.

File: src/functions/errorMessages.ts [hunk 13]
```diff
@@ -11,10 +11,19 @@ export type ErrorMessageMessage =
   | "Beginning with http(s):// is a must."
   | "The link is too long."
   | undefined;
+export type ErrorTagState = "assigned" | "empty" | "submit" | "wrong";
+export type ErrorTagMessage =
+  | "Tag already assigned"
+  | ""
```
Comment (erikvanek) [inline src/functions/errorMessages.ts]: Is an empty string an error message? I think it would make more sense to have an undefined at that moment instead of an empty string.

File: src/App.tsx (hunk 3 above)
Comment (brtsmrtn) [reply src/App.tsx]: True that

File: src/App.tsx (hunk 5 above)
Comment (brtsmrtn) [reply src/App.tsx]: Right!

File: src/components/Listing.tsx (hunk 6 above)
Comment (brtsmrtn) [reply src/components/Listing.tsx]: Great!

File: src/functions/errorMessages.ts (hunk 13 above)
Comment (brtsmrtn) [reply src/functions/errorMessages.ts]: Great, that makes absolute sense
